File: app/src/main/java/com/example/android/designpattern/java/singleton/Singleton.java
package com.example.android.designpattern.java.singleton;

public class Singleton {
    // Eager initialization
//    private static Singleton instance;
//    static {
//        if (instance == null) {
//            instance = new Singleton();
//        }
//    }
//
//    private Singleton() {}
//
//    public static Singleton getInstance() {
//        return instance;
//    }
    // Lazy initialization
    private static Singleton instance;

    private Singleton() {}

    public Singleton getInstance() {
        if (instance == null) {
            synchronized (Singleton.class) {
                if (instance == null) {
                    instance = new Singleton();
                }
            }
        }
        return instance;
    }
}

File: app/src/main/java/com/example/android/designpattern/kotlin/singleton/Singleton.kt
package com.example.android.designpattern.kotlin.singleton

class Singleton {
    // Eager initialization
//    companion object {
//        private lateinit var instance: Singleton
//
//        fun getInstance(): Singleton {
//            return instance
//        }
//    }
    // Lazy initialization
    companion object {
        private lateinit var instance: Singleton

        fun getInstance(): Singleton {
            if(instance == null) {
                synchronized(Singleton::class.java) {
                    if (instance == null) {
                        instance = Singleton()
                    }
                }
            }
            return instance
        }
    }
}
